### PR TITLE
Clears unacceptable date on input focus or calendar button click.

### LINKED ISSUE #828: Datepicker shows incorrect year in select when supplied out of range date
# Issue Description
If the date picker is supplied with a date outside the range of allowable dates, the year Select displays the incorrect year.

## Issue Type

- [ ] New Feature
- [ ] Enhancement
- [x] Bug
- [ ] Other

## Expected Behavior
Display the correct year value in the Select, and have the value disabled.

## Current Behavior
The invalid month is displayed in the month/day picker, while the Select shows the incorrect year. Attempting to change to the correct year is not possible until first changing to another year (onChange isn't occurring until that point).

## Steps to Reproduce
Supplied:
range -  September 18, **2011** to September 18, **2017**
selectedDate - September 18, **2009**

The result looks like:

wrong year

This is the month of September 2009, while the Select's say it is September 2011.

## Environment
* Component Version: terra-date-picker
* Browser Name and Version: 1.9.0
* Operating System and version (desktop or mobile): desktop

### PR DESCRIPTION
**Pull request by benbcai**
### Summary
Closes #828.

#### Changed Behavior
- When the default date is out of range or excluded, putting focus on the input field or clicking on the calendar button will clear that unacceptable default date as if no default date was provided.

### Deployment Links
For functional review go to the last two tests in these links and verify that the defaulted date/time is cleared when putting focus on the date or time input field or clicking on the calendar button to show the picker.
- https://terra-core-deployed-pr-974.herokuapp.com/static/#/tests/date-picker-tests
- https://terra-core-deployed-pr-974.herokuapp.com/static/#/tests/date-time-picker-tests

### CODE REVIEW

File: packages/terra-date-picker/src/DateInput.jsx [hunk 1]
```diff
@@ -15,10 +15,18 @@ const propTypes = {
    */
   inputAttributes: PropTypes.object,
   /**
+   * Indicates whether or not the picker should be displayed after rendering.
+   */
+  isForceShowPicker: PropTypes.bool,
```
Comment (JakeLaCombe) [inline packages/terra-date-picker/src/DateInput.jsx]: The wording of this seems a little odd. Maybe isPickerForceShown?

File: packages/terra-date-picker/src/DatePicker.jsx [hunk 2]
```diff
@@ -105,10 +113,27 @@ class DatePicker extends React.Component {
       selectedDate: DateUtil.createSafeDate(props.selectedDate),
     };
 
+    this.isDefaultDateAcceptable = true;
```
Comment (JakeLaCombe) [inline packages/terra-date-picker/src/DatePicker.jsx]: Minor:

In my opinion, it's usually better to assume the data is not acceptable until you verify that it is. I think this should default to false.

File: packages/terra-date-picker/src/DatePicker.jsx [hunk 3]
```diff
@@ -159,6 +184,44 @@ class DatePicker extends React.Component {
...
+
+  handleOnCalendarButtonClick(event, onClick) {
+    if (this.props.onCalendarButtonClick) {
+      this.props.onCalendarButtonClick(event);
+    }
+
+    if (!this.isDefaultDateAcceptable && !this.validateDefaultDate()) {
+      this.handleChange(null, event);
+    } else if (onClick) {
+      onClick();
```
Comment (JakeLaCombe) [inline packages/terra-date-picker/src/DatePicker.jsx]: perhaps we should pass the event to onClick?

File: packages/terra-date-picker/src/DatePicker.jsx [hunk 4]
```diff
@@ -159,6 +184,44 @@ class DatePicker extends React.Component {
...
+      this.isDefaultDateAcceptable = true;
+    }
+  }
+
+  validateDefaultDate() {
+    let iSAccepatable = true;
```
Comment (JakeLaCombe) [inline packages/terra-date-picker/src/DatePicker.jsx]: I'd keep the S lowercase (isAcceptable)

File: packages/terra-date-picker/tests/jest/DatePicker.test.jsx [hunk 5]
```diff
@@ -130,3 +130,23 @@ it('throws error on missing locale prop in Base', () => {
...
+      <DatePicker name="date-input" onCalendarButtonClick={handleOnCalendarButtonClick} utcOffset={0} />
+    </IntlProvider>,
+  );
+  expect(datePicker).toMatchSnapshot();
+});
+
+it('should render a date picker with onInputFocus', () => {
+  const handleOnInputFocus = () => {};
+  const datePicker = shallow(
+    <IntlProvider locale={locale} messages={messages}>
```
Comment (JakeLaCombe) [inline packages/terra-date-picker/tests/jest/DatePicker.test.jsx]: Is there a reason we aren't using the TerraI18nProvider here?

File: packages/terra-date-picker/src/DateUtil.js [hunk 6]
```diff
@@ -30,6 +30,42 @@ class DateUtil {
...
+    return true;
+  }
+
+  // Checks if a given date is one of the excluded dates.
+  static isDateExcluded(sourceDate, excludedDates) {
+    if (!sourceDate || !sourceDate.isValid()) {
+      return false;
+    }
+
+    const exludeMomentDates = DateUtil.filterInvalidDates(excludedDates);
```
Comment (JakeLaCombe) [inline packages/terra-date-picker/src/DateUtil.js]: excludeMomentDates*

File: packages/terra-date-picker/src/DateUtil.js [hunk 7]
```diff
@@ -30,6 +30,42 @@ class DateUtil {
...
+
+    if (!exludeMomentDates || exludeMomentDates.length <= 0) {
```
Comment (JakeLaCombe) [inline packages/terra-date-picker/src/DateUtil.js]: Do you need this if check? It looks like your for statement covers the case when this array is empty.

File: packages/terra-date-picker/tests/nightwatch/components/DatePickerDefaultDateExcluded.jsx [hunk 8]
```diff
@@ -0,0 +1,17 @@
+import React from 'react';
+import Base from 'terra-base';
+import DatePicker from '../../../lib/DatePicker';
+
+const locale = document.getElementsByTagName('html')[0].getAttribute('lang');
+
+const DatePickerDefaultDateExcluded = () => (
+  <Base locale={locale}>
```
Comment (JakeLaCombe) [inline packages/terra-date-picker/tests/nightwatch/components/DatePickerDefaultDateExcluded.jsx]: In some examples you use the IntlProvider and in these you use base. Can it be the same for all tests?

**Comment by mmalaker** [comment]
Functional verification completed. The date and time test examples display and function as expected when given an unacceptable date.

File: packages/terra-date-picker/src/DateInput.jsx (hunk 1 above)
Comment (kschuste) [reply packages/terra-date-picker/src/DateInput.jsx]: the prefix of 'is' is throwing me off for this name, I think just calling it forceShowPicker would work better

File: packages/terra-date-picker/src/DateUtil.js [hunk 9]
```diff
@@ -30,6 +30,42 @@ class DateUtil {
...
 
+  // Checks if a given date is out of the range between the start and end dates.
+  static isDateOutOfRange(sourceDate, startDate, endDate) {
+    if (!sourceDate || !sourceDate.isValid() ||
+      !startDate || !startDate.isValid() ||
+      !endDate || !endDate.isValid()) {
+      return false;
+    }
+
+    if (sourceDate.isBefore(startDate) && sourceDate.isAfter(endDate)) {
```
Comment (kschuste) [inline packages/terra-date-picker/src/DateUtil.js]: this should be an ||

the sourceDate can't both be before the start and after the end dates

File: packages/terra-date-picker/src/DatePicker.jsx (hunk 3 above)
Comment (benbcai) [reply packages/terra-date-picker/src/DatePicker.jsx]: This onClick is actually react-datepicker's [onInputClick](https://github.com/Hacker0x01/react-datepicker/blob/master/src/datepicker.jsx#L474) and it doesn't take any [parameters](https://github.com/Hacker0x01/react-datepicker/blob/master/src/datepicker.jsx#L326-L330).

File: packages/terra-date-picker/src/DateInput.jsx (hunk 1 above)
Comment (benbcai) [reply packages/terra-date-picker/src/DateInput.jsx]: I renamed to shouldShowPicker. https://github.com/cerner/terra-core/pull/974/commits/f120301502e3d390e72a669fce130604de572ae5

File: packages/terra-date-picker/src/DatePicker.jsx (hunk 2 above)
Comment (benbcai) [reply packages/terra-date-picker/src/DatePicker.jsx]: Updated: https://github.com/cerner/terra-core/pull/974/commits/f120301502e3d390e72a669fce130604de572ae5

File: packages/terra-date-picker/src/DatePicker.jsx (hunk 4 above)
Comment (benbcai) [reply packages/terra-date-picker/src/DatePicker.jsx]: Updated: https://github.com/cerner/terra-core/pull/974/commits/f120301502e3d390e72a669fce130604de572ae5

File: packages/terra-date-picker/tests/jest/DatePicker.test.jsx (hunk 5 above)
Comment (benbcai) [reply packages/terra-date-picker/tests/jest/DatePicker.test.jsx]: This is following the same way as the other components.

File: packages/terra-date-picker/src/DateUtil.js (hunk 6 above)
Comment (benbcai) [reply packages/terra-date-picker/src/DateUtil.js]: Updated: https://github.com/cerner/terra-core/pull/974/commits/f120301502e3d390e72a669fce130604de572ae5

File: packages/terra-date-picker/src/DateUtil.js (hunk 7 above)
Comment (benbcai) [reply packages/terra-date-picker/src/DateUtil.js]: Updated: https://github.com/cerner/terra-core/pull/974/commits/f120301502e3d390e72a669fce130604de572ae5

File: packages/terra-date-picker/tests/nightwatch/components/DatePickerDefaultDateExcluded.jsx (hunk 8 above)
Comment (benbcai) [reply packages/terra-date-picker/tests/nightwatch/components/DatePickerDefaultDateExcluded.jsx]: This is following the same way as the other components.

File: packages/terra-date-picker/src/DateUtil.js (hunk 9 above)
Comment (benbcai) [reply packages/terra-date-picker/src/DateUtil.js]: Updated: https://github.com/cerner/terra-core/commit/f120301502e3d390e72a669fce130604de572ae5

File: packages/terra-date-picker/src/DateInput.jsx [hunk 10]
```diff
@@ -104,12 +124,21 @@ class DatePickerInput extends React.Component {
     }
   }
 
+  handleOnInputFocus(event) {
+    if (this.props.onInputFocus) {
```
Comment (JakeLaCombe) [inline packages/terra-date-picker/src/DateInput.jsx]: This wrapper feels extraneous, as your onInputFocus is all that's done on focus. Can you remove this extraneous wrapper and just pass this.props.onInputFocus to the element?

Comment (benbcai) [reply packages/terra-date-picker/src/DateInput.jsx]: Updated. https://github.com/cerner/terra-core/pull/974/commits/f55edc6bf3657e8b9740cd65f5c159af25b21eb1

File: packages/terra-date-picker/src/DateInput.jsx [hunk 11]
```diff
@@ -78,14 +91,20 @@ class DatePickerInput extends React.Component {
...
+  handleOnButtonClick(event) {
     // The picker is about to display so request focus from the containing component (e.g. modal) if it has the focus trapped.
     if (this.props.requestFocus) {
       this.props.requestFocus();
     }
 
-    if (this.props.onClick) {
-      this.props.onClick();
+    if (this.props.onCalendarButtonClick) {
+      this.props.onCalendarButtonClick(event, this.props.onClick);
```
Comment (kschuste) [inline packages/terra-date-picker/src/DateInput.jsx]: want to add if check on if this.props.onClick is valid

File: packages/terra-date-picker/src/DatePicker.jsx (hunk 3 above)
Comment (emilyrohrbough) [reply packages/terra-date-picker/src/DatePicker.jsx]: ^ Can we add inline documentation indicate this onClick doesn't take parameters?

File: packages/terra-date-picker/src/DateUtil.js [hunk 12]
```diff
@@ -30,6 +30,40 @@ class DateUtil {
...
 
+  // Checks if a given date is out of the range between the start and end dates.
+  static isDateOutOfRange(sourceDate, startDate, endDate) {
+    if (!sourceDate || !sourceDate.isValid() ||
+      !startDate || !startDate.isValid() ||
+      !endDate || !endDate.isValid()) {
+      return false;
+    }
+
+    if (sourceDate.isSameOrAfter(startDate) && sourceDate.isSameOrBefore(endDate)) {
```
Comment (emilyrohrbough) [inline packages/terra-date-picker/src/DateUtil.js]: Are .isSameOrAfter & .isSameOrBefore moment functions?

File: packages/terra-date-time-picker/src/DateTimePicker.jsx [hunk 13]
```diff
@@ -264,6 +281,47 @@ class DateTimePicker extends React.Component {
...
+  handleOnInputFocus(event) {
+    if (!this.isDefaultDateAcceptable) {
+      this.dateValue = '';
+      this.timeValue = '';
+      this.handleChange(event, null);
+      this.isDefaultDateAcceptable = true;
+    }
+  }
+
+  handleOnCalendarButtonClick(event) {
```
Comment (emilyrohrbough) [inline packages/terra-date-time-picker/src/DateTimePicker.jsx]: Looking at these examples, should `handleOnCalendarButtonClick` and `handleOnInputFocus` actually be handled by the date-picker themselves? I'm just curious

File: packages/terra-date-picker/src/DatePicker.jsx (hunk 3 above)
Comment (benbcai) [reply packages/terra-date-picker/src/DatePicker.jsx]: Added inline comment: https://github.com/cerner/terra-core/pull/974/commits/619c1d70775ed668582319dfa0410181bc8e4e55

File: packages/terra-date-picker/src/DateInput.jsx (hunk 11 above)
Comment (benbcai) [reply packages/terra-date-picker/src/DateInput.jsx]: Updated: https://github.com/cerner/terra-core/pull/974/commits/619c1d70775ed668582319dfa0410181bc8e4e55

File: packages/terra-date-picker/src/DateUtil.js (hunk 12 above)
Comment (benbcai) [reply packages/terra-date-picker/src/DateUtil.js]: Yes, these are moment [functions](http://momentjs.com/docs/#/query/is-same-or-before/).

File: packages/terra-date-picker/src/DatePicker.jsx (hunk 3 above)
Comment (emilyrohrbough) [reply packages/terra-date-picker/src/DatePicker.jsx]: Thanks @benbcai

File: packages/terra-date-time-picker/src/DateTimePicker.jsx (hunk 13 above)
Comment (benbcai) [reply packages/terra-date-time-picker/src/DateTimePicker.jsx]: The date-picker and time-input do handle these events to a certain degree. When these events occur, the date-time-picker would also need to manage and update its states and handle the change.

**Pull request merged**